### ISSUE TITLE
Different outputs between igraph and leidenalg with self loops

### ISSUE THREAD

**Issue by Puumanamana** [igraph/python-igraph#331]
Make sure that these boxes are checked before submitting your issue -- thank you!

- [x] This issue is for the Python interface of igraph.
- [x] This issue is a bug report or a feature request, not a support question.

Hi,
I have recently discovered that the leiden clustering algorithm was included in python-igraph, and this is very helpful for me. It mostly produces very similar results as the `leidenalg` package. However, there seems to be a systematic difference between the 2 packages when vertices have self-edges. Here's an example:

```python
import igraph # v0.8.2
import leidenalg #  v0.8.2


def get_graph():
    g = igraph.Graph()
    g.add_vertices(['V0', 'V1', 'V2', 'V3'])
    g.add_edges([('V0', 'V1'), ('V2', 'V3')])

    return g

def print_igraph_clusters(g):
    clusters = g.community_leiden('CPM', resolution_parameter=0.1)
    print([cluster for cluster in clusters])

def print_leiden_clusters(g):
    # optimiser = leidenalg.Optimiser()
    # partition = leidenalg.CPMVertexPartition(cluster_graph, resolution_parameter=gamma)
    # optimiser.optimise_partition(partition)
    clusters = leidenalg.find_partition(
        g, leidenalg.CPMVertexPartition, resolution_parameter=0.1
    )

    print([cluster for cluster in clusters])

> g = get_graph()
# same output for both
> print_igraph_clusters(g)
[[0, 1], [2, 3]]
> print_leiden_clusters(g)
[[0, 1], [2, 3]]

> g.add_edges([(f'V{i}', f'V{i}') for i in range(4)])
# igraph produces singleton clusters while leiden produce the same output as before 
> print_igraph_clusters(g)
[[0], [1], [2], [3]]
> print_leiden_clusters(g)
[[0, 1], [2, 3]]
```

**Comment by ntamas** [comment]
Summon @vtraag :)

**Comment by vtraag** [comment]
This essentially goes back to the fact that self-loops are included twice in incident edges in `igraph`. I correct for this in `leidenalg`, but I don't correct for this in the `community_leiden` implementation. I should probably call `igraph_inclist_remove_duplicate` to remove those duplicates. 

It is good that the docs are now clear on this (after https://github.com/igraph/igraph/issues/1396). Perhaps we should also question whether there is ever a situation in which we would like self-loops to appear twice in incident edges / neighbor lists. I understand that for some definitions it may make sense to weight self-loops doubly (see https://github.com/igraph/igraph/issues/1193), but to really also have them twice in a list seems not so useful to me.

**Comment by vtraag** [comment]
To clarify the situation a bit further. When aggregating the graph, the self-loops are necessary in `leidenalg` to ensure that the quality of the singleton partition on the aggregated graph is identical to the quality of the original partition on the original graph. For that purpose, the weight of a self-loop must be counted only once, not twice. This is why that contribution is halved in `leidenalg`. In `community_leiden` we use a slightly different system, where we don't keep track of the self-loops in the aggregated graphs, but they are still present in the original graph of course.

**Comment by Puumanamana** [comment]
I'm not sure I fully understand, are you saying Leiden treats a vertex differently if he has a self-loop or not? Or do you mean that you need a self-loop on all edges anyway to make the singletons partition "quality" (CPM or modularity I guess) comparable to non-singleton partitions? 
In the general case, I'm not sure why self-loops would matter in a community detection algorithm.

**Comment by vtraag** [comment]
No, I just meant to clarify why the two implementations use a different approach. In general, the self-loops should not matter for determining the community with the maximum improvement indeed: moving a node moves along the self-loops as well. However, in the `leidenalg` implementation, the setup is more general so that all different kinds of objective functions can be maximised. Because of this, we explicitly need to include self-loops (if you are really interested I can explain the details). In the implementation in `igraph` we actually ignore self-loops.

But, now that I am thinking of it, there is actually a problem, because the self-loops are included when calculating the quality for the own community:

https://github.com/igraph/igraph/blob/ec82c3afcf3e0b4d2d14718f5f132c2e6b2a2d9f/src/community_leiden.c#L150

but the self-loops are not included when calculating the improvement for alternative communities:

https://github.com/igraph/igraph/blob/ec82c3afcf3e0b4d2d14718f5f132c2e6b2a2d9f/src/community_leiden.c#L153

because `VECTOR(edge_weights_per_cluster)[current_cluster]` is based on edges retrieved from the incident list, including the self-loops. We should therefore filter out any self-loops before running the algorithm.

@tamas, @szhorvat, also in view of the broader discussion about incident lists (cf. https://github.com/igraph/igraph/issues/1474), should we perhaps have a separate function `igraph_inclist_remove_self_loops`? The alternative is simple: it can also be done during the running of the algorithm (i.e. simply skip self-loops), but this is less efficient, since this check is then done more frequently than necessary.
